### ISSUE TITLE
external command causes SIGINT to be ignored within non-forking subshell

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#86]
On a ksh compiled with `-DSHOPT_SPAWN=0` to disable AST `spawnveg`(3) (see #79), running an external command in a non-forking subshell blocks SIGINT until that subshell is exited. If the subshell never exits, ksh must be killed from another shell using another signal.

This applies to both interactive shells and scripts.

Reproducer:
```
$ arch/*/bin/ksh -c '(/usr/bin/true; while :; do :; done); exit'
^C^C^C^C^C
```
(the extra 'exit' is to avoid the subshell being the last command, which would optimise it out when using `-c`)

**Comment by McDutchie** [comment]
This bug is not present in ksh2020, which has removed `spawnveg`.

**Comment by McDutchie** [comment]
This probably doesn't need to be a blocking bug now that #79 is fixed.

**Comment by McDutchie** [comment]
Actually, after the fix for #79, now this bug still occurs when compiled with `SHOPT_SPAWN` on, and if the shell is interactive and has job control on:
```
$ arch/*/bin/ksh -i -c '(/usr/bin/true; while :; do :; done); exit'
^C^C^C^C^C
```

**Comment by McDutchie** [comment]
It appears that the root cause is probably the fact that ksh only properly disables job control in non-forked/virtual subshells if they are command substitutions. I'm testing a fix for #89 and that fix makes this bug go away as well.

**Comment by McDutchie** [comment]
> It appears that the root cause is probably the fact that ksh only properly disables job control in non-forked/virtual subshells if they are command substitutions. I'm testing a fix for #89 and that fix makes this bug go away as well.

Of course it does; the #89 fix makes ksh use `sh_ntfork()` instead of `sh_fork()`, where the bug is. So that was a false lead.

**Comment by McDutchie** [comment]
This diff fixes it for me. For some reason I can't fathom, `sh_fork()` did not reset `shp->savesig` when in a virtual subshell, causing `SIGINT` to be ignored after running an external command using `sh_fork()` for the duration of the virtual subshell. The patch does not cause any regression test failures for me, on a ksh compiled with and without `-DSHOPT_SPAWN=0`.
```diff
diff --git a/src/cmd/ksh93/sh/xec.c b/src/cmd/ksh93/sh/xec.c
index a6ad507..baa8d6d 100644
--- a/src/cmd/ksh93/sh/xec.c
+++ b/src/cmd/ksh93/sh/xec.c
@@ -2951,13 +2951,10 @@ pid_t sh_fork(Shell_t *shp,int flags, int *jobid)
 	shp->savesig = -1;
 	while(_sh_fork(shp,parent=fork(),flags,jobid) < 0);
 	sh_stats(STAT_FORKS);
-	if(!shp->subshell)
-	{
-		sig = shp->savesig;
-		shp->savesig = 0;
-		if(sig>0)
-			kill(getpid(),sig);
-	}
+	sig = shp->savesig;
+	shp->savesig = 0;
+	if(sig>0)
+		kill(getpid(),sig);
 	job_fork(parent);
 	return(parent);
 }
```
